File: QytetetJava/src/modeloqytetet/TituloPropiedad.java
package modeloqytetet;
import java.util.ArrayList;
    /* 
     Esta clase representa
     el conjunto de características
     que se asocian a la propiedad
     de una casilla de tipo calle
     en qytetet
    */

public class TituloPropiedad {
    
// Campos privados asociados a la propiedad
    
        // Nombre de la calle
    private String nombre;
        // ¿Está hipotecada?
    private boolean hipotecada = false;
        // Su precio
    private int precioCompra;
        // Precio base por caer en la casilla
    private int alquilerBase;
        /*
           Revalorización de la propiedad entre
           su compra y venta 
        */
    private float factorRevalorizacion;
        // Valor base de su hipoteca
    private int hipotecaBase;
        // Precio de edificación
    private int precioEdificar;
        // Número de casas edificadas
    private int numCasas = 0;
        // Número de hoteles edificados
    private int numHoteles = 0;
    
        // Propietario de del titulo
    private Jugador propietario;
    
        // Constructor
    TituloPropiedad
        (String nombre, int precioCompra, int alquilerBase,
         float factorRevalorizacion, int hipotecaBase, 
         int precioEdificar)
        {
            this.nombre = nombre;
            this.precioCompra = precioCompra;
            this.alquilerBase = alquilerBase;
            this.factorRevalorizacion = factorRevalorizacion;
            this.hipotecaBase = hipotecaBase;
            this.precioEdificar = precioEdificar;
        }
        
    // Consultores

    String getNombre() {return nombre;}

    boolean getHipotecada() {return hipotecada;}

    int getPrecioCompra() {return precioCompra;}

    int getAlquilerBase() {return alquilerBase;}

    float getFactorRevalorizacion() {return factorRevalorizacion;}
    
    int getHipotecaBase() {return hipotecaBase;}

    int getPrecioEdificar() {return precioEdificar;}

    int getNumCasas() {return numCasas;}

    int getNumHoteles() {return numHoteles;}
    
    Jugador getPropietario () {return propietario;}
    
    boolean tengoPropietario () {return propietario != null;}
    
    boolean propietarioEncarcelado () {return propietario.getEncarcelado();}
    
        // Modificadores
    
    void setHipotecada (boolean hipotecada){
        this.hipotecada = hipotecada;
    }
    
    void setPropietario (Jugador propietario){
        this.propietario = propietario;
    }
    
    // Métodos que se usarán durante el transcurso del juego
    
    /*
        Incrementa el número de casas del título propiedad
    */
    
    void edificarCasa (){
        numCasas++;
    }
    
    /*
        Incrementa el número de hoteles del título propiedad
    */
    
    void edificarHotel (){
        numHoteles++;
    }
    
    /*
        Paga el alquiler del título
    */
    
    int pagarAlquiler(){
        int costeAlquiler = calcularCosteAlquiler();
        
        propietario.modificarSaldo(costeAlquiler);
        
        return costeAlquiler;
    
    }
    
    /*
        Calcula el coste del alquiler
    */
    
    int calcularCosteAlquiler(){
        return alquilerBase + (int) (numCasas * 0.5 + numHoteles * 2 );
    }

    /*
        Hipoteca la propiedad
    */
    
    int hipotecar(){
        setHipotecada(true);
        
        int costeHipoteca = calcularCosteHipotecar();
        
        return costeHipoteca;
    }
    
    /*
        Cancela la hipoteca
    */
    
    void cancelarHipoteca(){
        setHipotecada(false);
    }
    
    /*
        Calcula el coste de hipotecar la propiedad
    */
    
    int calcularCosteHipotecar(){
        return (int) (hipotecaBase + numCasas * 0.5 * hipotecaBase + numHoteles * hipotecaBase);
    }
    
    /*
        Calcula el precio de venta de la propiedad
    */
    
    int calcularPrecioVenta(){
        return (int) (precioCompra + (numCasas + numHoteles ) * precioEdificar * factorRevalorizacion);
    }
    
    // Método toString de la clase
    
    @Override
    public String toString() {
        String toReturn = "TituloPropiedad{" 
                + "nombre=" + nombre 
                + ", hipotecada=" + hipotecada 
                + ", precioCompra=" + precioCompra 
                + ", alquilerBase=" + alquilerBase 
                + ", factorRevalorizacion=" + factorRevalorizacion 
                + ", hipotecaBase=" + hipotecaBase 
                + ", precioEdificar=" + precioEdificar 
                + ", numCasas=" + numCasas 
                + ", numHoteles=" + numHoteles;
        
        // Solución a el problema de la navegabilidad en doble sentido
        
        if (propietario != null)
                toReturn += ", propietario=" + propietario.getNombre();
        
        return toReturn + '}';
    }
}

File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/ModeloQytetet/titulo_propiedad.rb
# encoding: utf-8

module ModeloQytetet

  class TituloPropiedad
    
    # Consultores y modificadores
    
    attr_reader:nombre
    attr_accessor:hipotecada,:propietario
    attr_reader:factor_revalorizacion
    attr_reader:alquiler_base
    attr_reader:precio_compra
    attr_reader:hipoteca_base
    attr_reader:precio_edificar
    attr_reader:num_hoteles
    attr_reader:num_casas
    
    # Constructor que crea un título de propiedad a partir de 5 argumentos
    
    def initialize (nombre, precio_compra, alquiler_base, factor_revalorizacion, 
          hipoteca_base,precio_edificar)
      @nombre = nombre
      @precio_compra = precio_compra
      @hipotecada = false
      @alquiler_base = alquiler_base
      @factor_revalorizacion = factor_revalorizacion
      @hipoteca_base = hipoteca_base
      @precio_edificar = precio_edificar
      @num_hoteles = 0
      @num_casas= 0
      @propietario
    end
    
    # Devuelve true solo si tiene propietario
    
    def tengo_propietario
      return !@propietario.nil?
    end
    
    # Devuelve true solo si el propietario está encarcelado
    
    def propietario_encarcelado
      return @propietario.encarcelado
    end
    
    # Paga el alquiler al jugador propietario de la propiedad
    
    def pagar_alquiler
      coste_alquiler = calcular_importe_alquiler
      @propietario.modificar_saldo(coste_alquiler)
      
      return coste_alquiler
    end
    
    # Método auxiliar para calcular el coste del alquiler de la propiedad 
    
    def calcular_importe_alquiler
      coste_alquiler = @alquiler_base + (@num_casas * 0.5 + @num_hoteles * 2).to_i
      
      return coste_alquiler
    end
    
    # Hipoteca la propiedad
    
    def hipotecar
      @hipotecada = true
      return calcular_coste_hipotecar
    end
    
    # Método auxiliar para calcular el coste de hipotecar la propiedad
    
    def calcular_coste_hipotecar
      return (@hipoteca_base + @num_casas * 0.5 * @hipoteca_base + @num_hoteles * @hipoteca_base).to_i
    end
    
    # Método auxiliar para calcular el coste de venta de la propiedad
    
    def calcular_precio_venta
      return (@precio_compra + (@num_casas + @num_hoteles) * @precio_edificar * @factor_revalorizacion).to_i
    end
    
    # Incrementa el número de casas edificadas
    
    def edificar_casa
      @num_casas += 1
    end
    
    # Incrementa el número de hoteles edificados
    
    def edificar_hotel
      @num_hoteles += 1
      @num_casas -= 4
    end
    
    # Cancela la hipoteca de la propiedad
    
    def cancelar_hipoteca
      @hipotecada = false
    end
    
    # toString()
    
    def to_s()
       
      to_return = "TituloPropiedad; #{@nombre} \nhipotecada: #{@hipotecada}" 
      to_return << "\nalquiler_base: #{@alquiler_base}\nfactor_revalorizacion: " 
      to_return << "#{@factor_revalorizacion}\nhipoteca_base: #{@hipoteca_base}" 
      to_return << "\nprecio_edificar: #{@precio_edificar}\nnum_hoteles: #{@num_hoteles}"
      to_return << "\nnum_casas: #{@num_casas}\nprecio_compra: #{@precio_compra}"
      
      if (!propietario.nil?)
        to_return << "\nPropietario: #{@propietario.nombre}"
      end
      
      to_return
    end
    
  end
end